internal/scheduler/schedulerobjects/nodetype.go

```diff
@@ -0,0 +1,93 @@
+package schedulerobjects
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/G-Research/armada/pkg/api"
+)
+
+type (
+	taintsFilterFunc func(*v1.Taint) bool
+	labelsFilterFunc func(key, value string) bool
+)
+
+func NewNodeTypeFromNode(node *v1.Node, wellKnownLabels map[string]string, taintsFilter taintsFilterFunc, labelsFilter labelsFilterFunc) *NodeType {
+	return NewNodeType(node.Spec.Taints, node.GetLabels(), wellKnownLabels, taintsFilter, labelsFilter)
+}
+
+func NewNodeTypeFromNodeInfo(nodeInfo *api.NodeInfo, wellKnownLabels map[string]string, taintsFilter taintsFilterFunc, labelsFilter labelsFilterFunc) *NodeType {
+	return NewNodeType(nodeInfo.GetTaints(), nodeInfo.GetLabels(), wellKnownLabels, taintsFilter, labelsFilter)
+}
+
+func NewNodeType(taints []v1.Taint, labels, wellKnownLabels map[string]string, taintsFilter taintsFilterFunc, labelsFilter labelsFilterFunc) *NodeType {
+	unsetWellKnownLabels := getFilteredLabels(wellKnownLabels, func(key, _ string) bool {
+		_, ok := labels[key]
+		return !ok
+	})
+	if unsetWellKnownLabels == nil {
+		unsetWellKnownLabels = make(map[string]string)
+	}
+	taints = getFilteredTaints(taints, taintsFilter)
+	labels = getFilteredLabels(labels, labelsFilter)
+	return &NodeType{
+		Id:                   nodeTypeIdFromTaintsAndLabels(taints, labels, unsetWellKnownLabels),
+		Taints:               taints,
+		Labels:               labels,
+		UnsetWellKnownLabels: unsetWellKnownLabels,
+	}
+}
+
+// nodeTypeIdFromTaintsAndLabels generates an id that is unique for each combination
+// of taints, labels, and unset labels. We insert "$" characters to achieve this,
+// since taints and labels are not allowed to contain this character.
+// See:
+// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+// https://man.archlinux.org/man/community/kubectl/kubectl-taint.1.en
+func nodeTypeIdFromTaintsAndLabels(taints []v1.Taint, labels, unsetWellKnownLabels map[string]string) string {
```

Comment:
ho bad does this end up being when you have say 10k with a thousand labels each?  I'm guessing  it's fine, but just thinking of the current case where we will be doing  this on every call?